Only unzip if the npm package is gzipped

MortenHoustonLudvigsen opened the pull request:
It turns out that not all packages downloaded from npm are actually gzipped (even though they are named `*.tgz`). Specifically, I have seen this with the `@types` scope introduced recently for use with typescript. In theses cases jspm install fails when trying to unzip:

```sh
jspm install npm:@types/jquery
     Updating registry cache...
     Downloading npm:@types/jquery@1.10.28

err  Error: incorrect header check
    at Zlib._handle.onerror (zlib.js:370:17)
```

This pull request resolves this by checking whether the downloaded package is actually gzipped, and only extracts the tar contents if not.

MortenHoustonLudvigsen commented:
I can also make a pull request against branch 0.17, if that would be better.

guybedford commented on lib/npm.js:
```diff
@@ -15,6 +16,42 @@ var Npmrc = require('./npmrc');
+    read(size) {
+      this.push(buffer);
+      this.push(null);
+    }
```
We need to still support Node 0.10 in jspm 0.16, so this should be the full object form I think.

MortenHoustonLudvigsen replied:
If I use [node-buffer-peek-stream](https://github.com/seangarner/node-buffer-peek-stream), this code can disappear :-)

guybedford replied:
Any ideas re performance implications of either method?
On Mon, 08 Aug 2016 at 13:33, Morten Houston Ludvigsen <
notifications@github.com> wrote:

> In lib/npm.js <https://github.com/jspm/npm/pull/147#discussion_r73860465>:
>
> > +          resolve(decoded);
> > +        }
> > +      });
> > +    });
> > +  } else {
> > +    return Promise.resolve(buffer);
> > +  }
> > +}
> > +
> > +// Create a readable stream from a buffer
> > +function bufferStream(buffer) {
> > +  return new stream.Readable({
> > +    read(size) {
> > +      this.push(buffer);
> > +      this.push(null);
> > +    }
>
> If I use node-buffer-peek-stream
> <https://github.com/seangarner/node-buffer-peek-stream>, this code can
> disappear :-)
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/jspm/npm/pull/147/files/38cd6cebb1ba586f001ccdce47afebb3a1030cf0#r73860465>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkiyoA15kp2lubQ6LWcOXVFZuKqtTJSks5qdxQCgaJpZM4Jeg1d>
> .
>

MortenHoustonLudvigsen replied:
Well, I have not done any measurements, but the way I have done it in this PR the whole stream is read into memory as buffers and then concatenated. This must have some impact on performance, even though I don't quite know how big the impact is.

Using [node-buffer-peek-stream](https://github.com/seangarner/node-buffer-peek-stream) I guess there will be a small overhead in peeking at the first few bytes and "stuffing" them back into the stream, but we won't have the entire stream in memory, so I am pretty confident that this would be the more performant choice.

guybedford replied:
Ok, that sounds like it would be a better approach then.

We have a large memory overhead in jspm anyway, so yes we shouldn't add to that by buffering all archives into memory before writing.

MortenHoustonLudvigsen replied:
OK - give me a bit, and I'll make the changes. :-)

guybedford commented on lib/npm.js:
```diff
@@ -376,21 +428,23 @@ NPMLocation.prototype = {
+            .then(gunzip)
+            .then(extractTarBuffer)
+            .then(resolve)
+            .catch(reject);
```
This seems counter to the streams philosophy to inject promises.

Is there not an approach that works directly on the streams?

guybedford replied:
I guess it may be worth timing a simple example before and after to see if this affects performance... if not, then I can't complain...

MortenHoustonLudvigsen replied:
The reason I used Promises is that I did not want to add dependencies, and I need a peek at the first few bytes to determine whether the stream is gzipped.
If you don't mind introducing a dependency I can use [node-buffer-peek-stream](https://github.com/seangarner/node-buffer-peek-stream) to achieve this instead, keeping the flow stream-based.

guybedford commented:
Thanks so much for the PR, this looks really great!

When we've finished review on this, getting it into 0.17 as well would definitely be worth doing.

MortenHoustonLudvigsen commented:
I'm glad you like it.

MortenHoustonLudvigsen commented:
Right, this is now implemented by peeking at the first few bytes of the download stream to check if the contents are gzipped. I reverted the first commit and made a third.
If you decide that this should be merged, I am happy to squash the commits first. :-)

guybedford commented:
Looks great, thanks!

guybedford commented:
If you're able to do a PR for 0.17 too please do.
